### PR TITLE
add filterRadars

### PR DESCRIPTION
**Pull request by KurtXiao**
Hi Devin, 

I added a new method _filterRadars to do the filtering. I just took out the radars that are in the range of any other radar so that no two radars would cover the same area, because we actually need to take out a lot of radars to increase the speed of the code. For the plotRadarImages method, when radars === null and we are plotting all radars, _filterRadars is applied. And when radars !== null and we are plotting the radars selected by user, _filterRadars is applied when the number of radars exceeds 5 so that basic level of precision is guaranteed for users without slowing down the code. Do you think this will work?

### CODE REVIEW

**Comment by dragazo** [comment]
@KurtXiao This has all the commits from the previous PR. You need to go to your fork's home page on github and [pull upstream changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork). After that, you should rebase this branch onto the updated branch and force push this branch.

Alternatively, you could do `git diff @^ >temp.diff` to save your changes, then switch to the updated master branch (see above) and do `git apply temp.diff` to re-apply the changes to the updated master branch.

After force pushing this branch, you should just see one (or two) commits here and there should no longer be merge conflicts (see below).

File: src/server/services/procedures/nexrad-radar/nexrad-radar.js [hunk 1]
```diff
@@ -268,6 +268,17 @@ NexradRadar._addHurricane = async function(radar, plot, radarPlot, settings) {
...
 };
 
+NexradRadar._filterRadars = async function(radars) {
+    for(let i = 0; i < radars.length; ++i) {
+        for(let j = i + 1; j < radars.length; ++j) {
+            if(this._getDistanceFromLatLonInKm(RADAR_LOCATIONS[radars[i]][0], RADAR_LOCATIONS[radars[i]][1], RADAR_LOCATIONS[radars[j]][0], RADAR_LOCATIONS[radars[j]][1]) <= RANGE) {
+                radars.splice(j, 1);
+            }
+        }
+    }
```
Comment (dragazo) [inline src/server/services/procedures/nexrad-radar/nexrad-radar.js:278]: When you remove item `j`, all the indices after that do down by `1`, but you `++j` regardless. So if `j=1`, you delete `1`, `2` becomes the new `1`, but you `++j` up to `2`, which skips over the original `2`. Modifying arrays during iteration is sometimes dangerous like that - this case can be fixed by iterating `j` backwards until it reaches `i`.

**Review by dragazo — COMMENTED** [review]
As long as the result still looks nice, this sounds like a good approach. Have you tested the speed? On my machine, the previous version took around 85 seconds on my machine but nearly 5 min on the server, so keep in mind the server is not super powerful :)

One thing I noticed: you check pairwise distance and remove one of them, but you always remove whichever comes second (`j` rather than `i`). What if `i` is off the map but still close to `j` (on the map)? It might be better to start by sorting in ascending distance from the middle of the map so that `j` is always further away than `i`.

File: src/server/services/procedures/nexrad-radar/nexrad-radar.js (hunk 1 above)
Comment (KurtXiao) [reply src/server/services/procedures/nexrad-radar/nexrad-radar.js]: Hi Devin. I just finished the revision and made a new commit. Now on my computer it runs 40s when the zoom level is 5 and 30s when the zoom level is 6.

**Comment by dragazo** [comment]
@KurtXiao I just tried out the filtering. It's definitely a major speedup, but it looks like the current version might be impacting the quality too much (here are some tests I ran: [img.zip](https://github.com/NetsBlox/NetsBlox/files/7879081/img.zip)). It might be that you just need to tune the pruning range to be smaller, which could vary with zoom level. We just need to make sure it looks good at all zoom levels.

Also I tried zoom level 1 and it looks like nothing showed up. When I ran `listRadars`, it returned only 4 items, so that might not be working for very low zoom levels.

**Comment by KurtXiao** [comment]
@dragazo Hi Devin. I just finished the revision. I ran some tests [tests.zip](https://github.com/NetsBlox/NetsBlox/files/7884927/tests.zip) and I think the results look fine. There is one additional change I made. When we are sorting the radars in _filterRadars, I am sorting them in descending distance from the center of the map. I think because we are dealing with hurricanes, radars closer to the shore should be more informative, which are the radars away from the center. I feel like doing so might improve the quality of the images.

**Comment by dragazo** [comment]
@KurtXiao I just tested the most recent version and they all looked good with new filtering. The speedup is still pretty good - I got around 2x on the ones I tested.

The big slowdown is still the parsing library. There are some drastic things we can do to solve that (like rewriting that part in c++/rust), but that would be a big undertaking and we don't even know what the potential speedup would be until it's done. So best to avoid that, though if you happen to find such a version, we could talk about how to create node bindings to use it from js (rust would be easier for that part).

The only issue with the PR right now is that you included the `src/browser` submodule in the changes. You can revert that by doing `git checkout 225e031 src/browser` and committing,

**Comment by KurtXiao** [comment]
@dragazo I reverted back to d12fb14, whose src/browser is the same as that for 8f88fef. I will keep an eye on the parsing library!

**Comment by brollb** [comment]
Thanks, @dragazo and @KurtXiao!

**Pull request merged**